File: python/aoc2023/day04.py
from os.path import join

from aoc2023 import DATA_DIR


class Card:
    def __init__(
        self, card_num: int, winning_numbers: set[int], numbers: set[int]
    ) -> None:
        self.card_num = card_num
        self.winning_numbers = winning_numbers
        self.numbers = numbers
        self.num_matching = len(self.winning_numbers.intersection(self.numbers))
        self.points = 0 if self.num_matching == 0 else 1 << (self.num_matching - 1)


def read_input(filename: str) -> list[Card]:
    cards: list[Card] = []

    with open(join(DATA_DIR, filename), "r") as fh:
        for line in fh.read().strip().split("\n"):
            card, rest = line.split(":")
            left, right = rest.split("|")

            card_num = int(card.split()[1])
            winning_numbers = {int(num) for num in left.split() if num.strip() != ""}
            numbers = {int(num) for num in right.split() if num.strip() != ""}
            cards.append(Card(card_num, winning_numbers, numbers))

    return cards


def part1(filename: str) -> int:
    cards = read_input(filename)
    return sum([card.points for card in cards])


def part2(filename: str) -> int:
    cards = read_input(filename)

    num_cards = [1] * len(cards)
    for idx, card in enumerate(cards):
        for offset in range(card.num_matching):
            if idx + offset < len(cards):
                num_cards[idx + offset + 1] += num_cards[idx]

    return sum(num_cards)


if __name__ == "__main__":
    assert part1("day04_sample.txt") == 13
    assert part1("day04.txt") == 23678

    assert part2("day04_sample.txt") == 30
    assert part2("day04.txt") == 15455663

File: rust/src/day04.rs
use std::{collections::HashSet, fs, hash::Hash, num::ParseIntError};

use pest::{iterators::Pairs, Parser};
use pest_derive::Parser;

const MAX_CARD_ID: usize = 202;

#[derive(Parser)]
#[grammar = "day04.pest"] // relative to src
struct CardParser;

fn extract_numbers(pair_iter: &mut Pairs<Rule>) -> Vec<i32> {
    let numbers: Vec<i32> = pair_iter
        .next()
        .expect("no winning numbers")
        .into_inner()
        .map(|v| v.as_str().parse::<i32>())
        .flatten()
        .collect();
    numbers
}

pub fn part1(filename: &str) -> Result<usize, ParseIntError> {
    let unparsed = fs::read_to_string(filename).expect("cannot read file");
    let mut pairs = CardParser::parse(Rule::cards, &unparsed)
        .expect("cannot parse file")
        .into_iter();

    let mut sum: usize = 0;
    while let Some(item) = pairs.next() {
        if item.as_str() == "" {
            continue;
        }

        let _idx = item.as_str().parse::<usize>()?;

        let winning_numbers: HashSet<i32> = extract_numbers(&mut pairs).into_iter().collect();
        let card_numbers: HashSet<i32> = extract_numbers(&mut pairs).into_iter().collect();
        let num_winners = winning_numbers.intersection(&card_numbers).count();

        let score = match num_winners {
            0 => 0,
            1 => 1,
            c => 2_usize.pow((c as u32) - 1),
        };
        sum += score;
    }

    Ok(sum)
}

pub fn part2(filename: &str) -> Result<usize, ParseIntError> {
    let unparsed = fs::read_to_string(filename).expect("cannot read file");
    let mut pairs = CardParser::parse(Rule::cards, &unparsed)
        .expect("cannot parse file")
        .into_iter();

    let mut sum: usize = 0;
    let mut num_cards: [usize; MAX_CARD_ID + 1] = [0; MAX_CARD_ID + 1];

    while let Some(item) = pairs.next() {
        if item.as_str() == "" {
            continue;
        }

        let idx = item.as_str().parse::<usize>()?;
        num_cards[idx] += 1;

        let current_num_cards = num_cards[idx];

        let winning_numbers: HashSet<i32> = extract_numbers(&mut pairs).into_iter().collect();
        let card_numbers: HashSet<i32> = extract_numbers(&mut pairs).into_iter().collect();
        let num_winners = winning_numbers.intersection(&card_numbers).count();

        for i in (idx + 1)..(idx + 1 + num_winners) {
            num_cards[i] += current_num_cards;
        }

        sum += current_num_cards;
    }

    println!("NUM_CARDS: {num_cards:?}");

    Ok(sum)
}

#[cfg(test)]
mod tests {
    use super::{part1, part2};

    #[test]
    fn test_part1() {
        assert!(part1("../data/day04_sample.txt") == Ok(13));
        assert!(part1("../data/day04.txt") == Ok(23678));
    }

    #[test]
    fn test_part2() {
        assert!(part2("../data/day04_sample.txt") == Ok(30));
        assert!(part2("../data/day04.txt") == Ok(15455663));
    }
}
